smart_contract/executor/src/permission.rs

```diff
@@ -3,27 +3,49 @@
 use alloc::borrow::ToOwned as _;
 
 use iroha_schema::IntoSchema;
-use iroha_smart_contract::{data_model::permission::Permission, QueryOutputCursor};
+use iroha_smart_contract::{data_model::JsonString, QueryOutputCursor};
 use iroha_smart_contract_utils::debug::DebugExpectExt as _;
 use serde::{de::DeserializeOwned, Serialize};
 
-use crate::{data_model::prelude::*, prelude::*};
+use crate::{prelude::*, TryFromDataModelObjectError};
 
-/// [`Token`] trait is used to check if the token is owned by the account.
-pub trait Token:
+/// Is used to check if the permission token is owned by the account.
+pub trait Permission:
     Serialize + DeserializeOwned + IntoSchema + PartialEq<Self> + ValidateGrantRevoke
-where
-    for<'a> Self: TryFrom<&'a Permission, Error = PermissionConversionError>,
 {
-    /// Return name of this permission token
-    fn name() -> Name {
-        <Self as iroha_schema::IntoSchema>::type_name()
-            .parse()
-            .dbg_expect("Failed to parse permission token as `Name`")
+    /// Check if the account owns this token
+    fn is_owned_by(&self, account_id: &AccountId) -> bool;
+
+    /// Permission id, according to [`IntoSchema`].
+    fn id() -> PermissionId {
+        PermissionId::new(
+            <Self as iroha_schema::IntoSchema>::type_name()
+                .parse()
+                .dbg_expect("Failed to parse permission id as `Name`"),
+        )
     }
 
-    /// Check if token is owned by the account
-    fn is_owned_by(&self, account_id: &AccountId) -> bool;
+    /// Try to convert from [`PermissionObject`]
+    /// # Errors
+    /// See [`TryFromDataModelObjectError`]
+    fn try_from_object(object: &PermissionObject) -> Result<Self, TryFromDataModelObjectError> {
+        if *object.id() != <Self as Permission>::id() {
+            return Err(TryFromDataModelObjectError::Id(object.id().name().clone()));
+        }
+        object
+            .payload()
+            .deserialize()
+            .map_err(TryFromDataModelObjectError::Deserialize)
+    }
+
+    /// Convert into [`PermissionObject`]
+    fn to_object(&self) -> PermissionObject {
+        PermissionObject::new(
+            <Self as Permission>::id(),
+            JsonString::serialize(&self)
+                .expect("failed to serialize concrete data model entity; this is a bug"),
+        )
+    }
```

Comment:
what was wrong with having implemented `TryFrom` and `From`?

Reply:
🤔 good question

Reply:
Ah, yes. Because:

```rust
impl<T: Parameter> TryFrom<&ParameterObject> for T {
//                 ^^^^^^^^^^^^^^^^^^^^^^^^^
//                 Only traits defined in the current crate can be implemented for arbitrary types
    type Error = TryFromDataModelObjectError;

    fn try_from(value: &ParameterObject) -> core::result::Result<Self, Self::Error> {
        todo!()
    }
}
```

Reply:
you can't make a blanket implementation, but you can derive for each type on `#[derive(PermissionTrait)]`

Reply:
Then we also need to introduce a derive macro for `ParameterTrait`, and I wouldn't want it just for the sake of this conversion.